### ISSUE TITLE
400 Bad request when displaying image using JWT

### ISSUE THREAD

**Issue by a1r93** [upload-js/upload-js#12]
### Version
"@bytescale/sdk": "^3.34.0"

### Description

I am trying to handle authentication to upload and fetch the files, but the fetching of the file doesn't seem to be working as a I get a 400 bad request when trying to display the image with `<img src={url} />`

We start by authenticating using the AuthManager

```typescript
await bytescale.AuthManager.beginAuthSession({
  accountId: "ACCOUNT_ID",
  authUrl: "AUTHENTICATION_URL",
  authHeaders: () => Promise.resolve({}),
});
```

We then upload an image using the upload manager

```typescript
const uploadManager = new bytescale.UploadManager({
  apiKey: publicKey,
})

const response = await uploadManager.upload({
  data: file,
  path: {
    folderPath: "/",
    fileName: getFileName(),
  },
});
```

This response contains the fileUrl, which we directly use to display the image:

```typescript
function Image({ response }: { response: { fileUrl: string } }) {
  if (!fileUrl) return null;

  return <img src={fileUrl} alt="test" />
}
```

### Current result

The upload works as expected, I can indeed find my image in the Bytescale dashboard. The upload is protected by JWT so I can confirm that the JWT authentication works as expected.

The GET image request does not work, I receive the following response:

![321940478-486f44d6-5122-4fda-9879-a402d6309730](https://github.com/bytescale/bytescale-javascript-sdk/assets/3638917/ea12d5fb-cf54-4ae4-b9f7-340517ef71b4)

Note that we can see the JWT is indeed used to get the image through the cookie.

The response tab displays the following message: "Failed to load response data: No data found for resource with given identifier"

### Expected result

The image is displayed after it got uploaded

Am I maybe missing something?

**Comment by ljwagerfield** [comment]
Hi @a1r93,

Thank you for reporting this!

tl;dr We recommend using service workers for JWT-based authentication, as cookie-based auth is becoming less reliable as browsers evolve. For detailed guidance, please see:

[BeginAuthSession - Bytescale SDK](https://www.bytescale.com/docs/sdks/javascript#BeginAuthSession)
[BeginAuthSessionParams - Bytescale](https://www.bytescale.com/docs/types/BeginAuthSessionParams#serviceWorkerScript)

The response you're receiving is a Bytescale CDN error; these are returned as JSON and usually contain helpful information that allows you to self-diagnose the issue. However, you're unable to see the response body of this error due to some quirks with how DevTools works with regards to CORS, which I'll describe below:

Under certain conditions, it appears that DevTools retrospectively treats requests as CORS requests when attempting to display response bodies in the network tab. Since the original request in this case was triggered by an `<img>` element and no `crossorigin` attribute was present, the request was not made as a CORS request. This means the browser didn't send the Origin request header. Consequently, the response from Bytescale lacks specific CORS-related headers like 'Access-Control-Allow-Origin' (ACAO), which rely on the Origin header. Therefore, the preview in DevTools fails, displaying the error "Failed to load response data: No data found for resource with given identifier".

To navigate around this issue and to ensure better compatibility across all browsers, especially those that block third-party cookies like Safari, using service workers for handling JWT authentication is the way to go. Service workers ensure that necessary credentials are included in every request, and they can help manage the Origin headers required for CORS compliance more reliably.

If you choose to continue using cookies and need to ensure that CORS policies are correctly handled, you could explicitly set the crossorigin attribute to "use-credentials" on your `<img>` tags. This ensures that the Origin header is included in your requests, thereby facilitating the server in sending the appropriate Access-Control-Allow-Origin and Access-Control-Allow-Credentials response headers, which will allow things like Bytescale CDN errors to show up correctly in your DevTools.

Let me know if this helps or if you have any more questions!

**Comment by a1r93** [comment]
Thank you for the quick response, adding a service worker did indeed fix the problem. But now I don't seem to be able to display the image I uploaded:

- I receive a 404 not found when using the url received in the upload response, the url looks like follows: https://upcdn.io/ACCOUNT_ID/raw/file_name.png
- When trying to display the image using https://upcdn.io/ACCOUNT_ID/image/file_name.png instead, I receive a 403 Forbidden response

In the JWT we give the following authorisations, do you see anything wrong with it?

```typescript
pathPermissions: [
        {
          match: {
            path: "/",
            scope: "Grandchildren+",
          },
          permissions: {
            read: {
              file: {
                downloadFile: ["*"],
                getFileDetails: true,
              },
              folder: {
                getFolderDescription: true,
                getFolderPublicPermissions: true,
                getFolderStorageLayer: true,
                listFolderChildren: true,
              },
            },
            write: {
              file: {
                createFile: true,
                deleteFile: true,
                overwriteFile: true,
              },
              folder: {
                createVirtualFolder: true,
                deleteVirtualFolder: true,
                setFolderDescription: true,
                setFolderPublicPermissions: true,
                setFolderStorageLayer: true,
              },
            },
          },
        },
      ]
```

I checked in the bytescale dashboard and the file does exist under the root folder /file_name.png. Thanks!

**Comment by ljwagerfield** [comment]
The permissions in your JWT are sufficient for downloading the file (and are perhaps too permissive, as they allow full read/write access to your entire account).

It's quite likely a timing issue; i.e. the service worker is being registered after the image is being loaded.

Let's close this issue and resume in a private support ticket, so we can discuss request logs and arrange a screenshare 👍
